pkg/services/ngalert/state/historian/core.go

```diff
@@ -12,6 +13,8 @@ import (
 	history_model "github.com/grafana/grafana/pkg/services/ngalert/state/historian/model"
 )
 
+const StateHistoryWriteTimeout = 30 * time.Second
```

Comment:
30 secs can be not enough when there is a connection starvation. I think we should make it higher and also propagate from config file.

Reply:
during the synthetic testing with 20k of rules. See description of PR https://github.com/grafana/grafana/pull/61201, there are graphs at the bottom of the PR description that show that with 25k rules the blocked time regularly went up to 11 days, and if you divide it by the number of rules (i.e. go-routines) you will get about 38s).

Also, here is a [link](https://ops.grafana-ops.net/goto/nIKK8wfVg?orgId=1) to some traces with a span `open session` and duration greater than the 30s. In our environment, it's a fairly rare thing but on-premise it can be worse. 

> why this would cause the connection to starve

this will not cause starvation but can be affected by it.


I think the hardcoded relatively low timeout is not the best solution because it does not consider the environment where the application can run. At least, not in this case. 
1. The purpose of the timeout is to make sure that spawn go-routines do not run indefinitely. The results of this go-routine are not used anywhere down the workflow.
2. The most probable cause of the timeout is that the go routine gets blocked on waiting for a connection to be 
available in the pool. That operation does not have any timeouts, and therefore go routine can wait indefinitely. 
3. The result of the low timeout is data-loss. The result of the high timeout is memory and CPU consumption, worsening connection starvation. I think the latter is a less bad outcome. 

Therefore, I think that ideally, we should make this timeout configurable and keep the default 30s. Otherwise, increase it to a pretty high duration, say 3-5 minutes, to reduce the chance of data loss to a negligible low value.

Reply:
Prior to this change, the 30 seconds was only the `http.Client`'s timeout for _requests only_.

So it stands to reason that the timeout for the full goroutine should be greater than that number, since that includes the request plus other work.

I'll go with 1 minute as a happy medium, we can monitor and adjust over time.

Reply:
IMO that's enough for parity in this PR, but in the future... I honestly don't see much of a reason to have two timeouts anyway. WDYT about dropping the timeout on the `http.Client` entirely, and just using one big timeout for the write job?